### PR TITLE
Add dataloader optimization to decrease the data transfer overhead from CPU to GPU

### PR DESCRIPTION
**Pull request by chenrui17**
## Summary

- Add data-pipeline performance optimizations for Alpamayo SFT recipes, including dataloader worker settings, zip feature cache, collate processor cache, TF32, and cuDNN benchmark toggles.
- Add reproducible benchmark scripts and step-time callback for `alpamayo1_sft` and `alpamayo1_5_sft`.
- Include benchmark reports with step 5-20 average wall-clock step time as the stable performance metric.

### Alpamayo-1.5 SFT

- Baseline: 41.354s/step
- Dataloader workers: 20.040s/step, 2.06x speedup
- Zip + collate cache: 12.032s/step, 3.44x speedup
- TF32 + cuDNN benchmark: 12.021s/step, 3.44x speedup

### Alpamayo-1 SFT

No-CoC:
- Baseline: 50.109s/step
- Final optimized: 10.092s/step, 4.97x speedup

CoC:
- Baseline: 25.607s/step
- Final optimized: 10.308s/step, 2.48x speedup

Note: CoC uses the reasoning mini clip index, so CoC and no-CoC baselines should be compared within their own groups rather than directly against each other.

## Test Plan

- Ran Alpamayo-1.5 SFT 4-way benchmark on 8x NVIDIA H20 using `Alpamayo-1.5-10B-A1-format`.
- Ran Alpamayo-1 SFT no-CoC and CoC 2x4 benchmark matrix on 8x NVIDIA H20 using `Alpamayo-R1-10B`.
- Verified each run completed 20 training steps and reported the stable average over steps 5-20.
- Confirmed benchmark helper files pass lint checks.

### CODE REVIEW

File: recipes/alpamayo1_5_sft/benchmark/run_benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (super-anova) [inline recipes/alpamayo1_5_sft/benchmark/run_benchmarks.py:1]: Could you add a README to inform users on how to run and read these benchmarks?

File: recipes/alpamayo1_5_sft/performance/collate_cache.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Reuse a single QwenProcessor for collate to avoid per-batch reconstruction."""
+
+from __future__ import annotations
+
+from typing import Any
+
+from alpamayo_r1.processor.qwen_processor import QwenProcessor
```
Comment (super-anova) [inline recipes/alpamayo1_5_sft/performance/collate_cache.py]: should this import from `alpamayo.processor.qwen_processor` instead? same as what you did for `recipes/alpamayo1_sft/performance/collate_cache.py`

Comment (chenrui17) [reply recipes/alpamayo1_5_sft/performance/collate_cache.py]: Yes, you are right, fixed it.

**Comment by super-anova** [comment]
Thanks! PR mostly looks good to me. Will take another pass tomorrow and approve if all looks good!

File: recipes/alpamayo1_5_sft/performance/apply_optimizations.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+import torch
+
+
+def apply_runtime_optimizations(cfg: dict[str, Any]) -> dict[str, Any]:
+    """Apply global PyTorch / CUDA knobs before model construction."""
+    applied: dict[str, Any] = {}
+    opt = cfg.get("optimizations") or {}
+
+    if opt.get("tf32", False):
```
Comment (super-anova) [inline recipes/alpamayo1_5_sft/performance/apply_optimizations.py]: is this behavior correct? if users set `tf32` to false, the following lines inside the if won't get executed?

File: recipes/alpamayo1_5_sft/benchmark/run_benchmarks.py (hunk 1 above)
Comment (chenrui17) [reply recipes/alpamayo1_5_sft/benchmark/run_benchmarks.py:1]: added

File: recipes/alpamayo1_5_sft/performance/apply_optimizations.py (hunk 3 above)
Comment (chenrui17) [reply recipes/alpamayo1_5_sft/performance/apply_optimizations.py]: fixed it and make it more reasonable

**Review by super-anova — APPROVED** [review]

File: recipes/alpamayo1_5_sft/performance/collate_cache.py (hunk 2 above)
Comment (super-anova) [reply recipes/alpamayo1_5_sft/performance/collate_cache.py]: Thanks!
